library/core/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java

```diff
@@ -1277,17 +1282,20 @@ private boolean readSample(ExtractorInput input) throws IOException, Interrupted
         }
       }
     } else {
-      int sampleHeaderSize = 0;
-      if (MimeTypes.AUDIO_AC4.equals(track.format.sampleMimeType)) {
+      if (MimeTypes.AUDIO_AC4.equals(track.format.sampleMimeType) && !isAc4HeaderAdded) {
```

Comment:
I didn't put this part of code to "if (parserState == STATE_READING_SAMPLE_START)". Instead, I set "isAc4HeaderAdded" in "if (parserState == STATE_READING_SAMPLE_START)" branch. If there is an exception in input reading, AC-4 header will not be added again.
I didn't put this part of code to if branch because "output" is not available at that time. And I don't want to change your original code a lot.